arch/powerpc/platforms/pseries/pci.c

```diff
@@ -54,25 +54,176 @@ void pcibios_name_device(struct pci_dev *dev)
 			}
 		}
 	}
-}   
+}
 DECLARE_PCI_FIXUP_HEADER(PCI_ANY_ID, PCI_ANY_ID, pcibios_name_device);
 #endif
-
 #ifdef CONFIG_PCI_IOV
+#define MAX_VFS_FOR_MAP_PE 256
+struct pe_map_bar_entry {
+	__be64     bar;       ///< Input:  Virtual Function BAR
+	__be16     rid;       ///< Input:  Virtual Function Router ID
+	__be16     pe_num;    ///< Output: Virtual Function PE Number
+	__be32     reserved;  ///< Reserved Space
+};
+
+int pseries_send_map_pe(struct pci_dev *pdev,
+			       u16 num_vfs,
+			       struct pe_map_bar_entry *vf_pe_array)
+{
+	struct pci_dn *pdn;
+	int   rc;
+	unsigned long buid, addr;
+	int ibm_map_pes = rtas_token("ibm,open-sriov-map-pe-number");
+
+	if (ibm_map_pes == RTAS_UNKNOWN_SERVICE)
+		return -EINVAL;
+
+	pdn = pci_get_pdn(pdev);
+	addr = rtas_config_addr(pdn->busno, pdn->devfn, 0);
+	buid = pdn->phb->buid;
+	spin_lock(&rtas_data_buf_lock);
+	memcpy(rtas_data_buf, vf_pe_array,
+	       num_vfs*sizeof(struct pe_map_bar_entry));
```

Comment:
I believe your size here should be RTAS_DATA_BUF_SIZE. Your rtas_data_buf is of size RTAS_DATA_BUF_SIZE and your vf_pe_array is also of that size. Thus if you are using the size of num_vfs*sizeof(struct pe_map_bar_entry) it can cause underflow or overflow.

Reply:
will do.